engine/runtime/src/main/scala/org/enso/compiler/pass/analyse/GatherDiagnostics.scala

```diff
@@ -67,10 +67,53 @@ case object GatherDiagnostics extends IRPass {
       case x                  => x.diagnostics.toList
     }.flatten
     DiagnosticsMeta(
-      diagnostics.distinctBy(d => (d.location, d.message))
+      diagnostics.distinctBy(d => new DiagnosticKeys(d))
     )
   }
 
+  final private class DiagnosticKeys(private val diagnostic: IR.Diagnostic) {
+
+    /** Equals is based on type of diagnostic, its location and its diagnostic keys.
+      */
+    override def equals(any: Any): Boolean = {
+      any match {
+        case other: DiagnosticKeys => {
+          if (diagnostic.getClass() != other.diagnostic.getClass()) {
+            return false
```

Comment:
Explicit return statements is highly discouraged in Scala

Reply:
@JaroslavTulach just to illustrate I (personally) find this equivalent FP code easier to parse than trying to understand multiple return statements. results may vary.
```
          (diagnostic.getClass() == other.diagnostic.getClass()) &&
          (diagnostic.location == other.diagnostic.location) &&
          (diagnostic
            .diagnosticKeys()
            .zipAll(other.diagnostic.diagnosticKeys(), null, null)
            .forall { case (e1, e2) => e1 != null && e2 != null && e1 == e2 })
```

Reply:
Oh, no, please don't force me to use `zipAll`! But I try to restrict my too imperative style in Scala. I hope [0eb4119](https://github.com/enso-org/enso/pull/3465/commits/0eb41196ea307d467be9fd138cc44cbc7639a178) is a sufficient compromise.